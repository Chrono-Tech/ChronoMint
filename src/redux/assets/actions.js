import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
