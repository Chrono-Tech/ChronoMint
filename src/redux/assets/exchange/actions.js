import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
