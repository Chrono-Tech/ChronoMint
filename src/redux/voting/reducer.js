import Immutable from 'immutable'
import VotingMainModel from '../../models/voting/VotingMainModel'

export const POLLS_VOTE_LIMIT = 'voting/POLLS_LIMIT'
export const POLLS_LOAD = 'voting/POLLS_LOAD'
export const POLLS_LIST = 'voting/POLLS_LIST'
export const POLLS_CREATE = 'voting/POLLS_CREATE'
export const POLLS_REMOVE = 'voting/POLLS_REMOVE'
export const POLLS_REMOVE_STUB = 'voting/POLLS_REMOVE_STUB'
export const POLLS_UPDATE = 'voting/POLLS_UPDATE'

export const VOTING_POLLS_COUNT = 'voting/POLLS_COUNT'

const initialState = new VotingMainModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case POLLS_VOTE_LIMIT:
      return state.voteLimitInTIME(action.voteLimitInTIME)
    case POLLS_LOAD:
      return state.isFetching(true)
    case POLLS_LIST:
      return state
        .isFetching(false)
        .isFetched(true)
        .list(new Immutable.Map(action.list))
    case POLLS_CREATE:
      return state
        .list(state.list().set(action.poll.poll().id(), action.poll))
    case POLLS_REMOVE_STUB:
      return state
        .list(
          state.list()
            .filter((poll) => {
              const hash = poll.transactionHash()
              return hash === null || hash !== action.transactionHash
            }),
        )
    case POLLS_UPDATE:
      return state
        .list(state.list().set(action.poll.poll().id(), action.poll))
    case POLLS_REMOVE:
      return state
        .list(state.list().delete(action.id))
    case VOTING_POLLS_COUNT:
      return state
        .pollsCount(action.count || state.pollsCount())
        .activePollsCount(action.activeCount || state.activePollsCount())
    default:
      return state
  }
}
