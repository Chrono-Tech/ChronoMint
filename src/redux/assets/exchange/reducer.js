import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.EXCHANGE_MANAGER_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
