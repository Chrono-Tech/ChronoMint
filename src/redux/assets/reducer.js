import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.ASSETS_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
