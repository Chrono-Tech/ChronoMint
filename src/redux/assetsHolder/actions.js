import { DUCK_SESSION } from '@/redux/session/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import TIMEHolderDAO from 'dao/TIMEHolderDAO'
import Amount from 'models/Amount'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TokenModel from 'models/tokens/TokenModel'
import { DUCK_TOKENS } from 'redux/tokens/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'

export const DUCK_ASSETS_HOLDER = 'assetsHolder'

export const TIME_HOLDER_INIT = 'timeHolder/INIT'
export const TIME_HOLDER_TIME_ADDRESS = 'timeHolder/timeAddress'
export const TIME_HOLDER_WALLET_ADDRESS = 'timeHolder/timeHolderWalletAddress'
export const TIME_HOLDER_ADDRESS = 'timeHolder/timeHolderAddress'

let timeHolderDAO: TIMEHolderDAO = null

const subscribeOnTokens = (token: TokenModel) => async (dispatch, getState) => {
  if (token.symbol() !== 'TIME') {
    return
  }

  dispatch({ type: TIME_HOLDER_TIME_ADDRESS, address: token.address() })

  // init
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: TIME_HOLDER_ADDRESS, address: timeHolderAddress })
  dispatch({ type: TIME_HOLDER_WALLET_ADDRESS, address: timeHolderWalletAddress })

  // TODO @dkchv: review again
  const contractNames = {}
  contractNames[ timeHolderAddress ] = 'TIME Holder'
  ApprovalNoticeModel.setContractNames(contractNames)

  // update allowance
  const timeDAO = tokenService.getDAO(token.id())
  const { account } = getState().get(DUCK_SESSION)
  const [ timeHolderAllowance, timeHolderWalletAllowance ] = await Promise.all([
    timeDAO.getAccountAllowance(timeHolderAddress, account),
    timeDAO.getAccountAllowance(timeHolderWalletAddress, account),
  ])

  console.log('--assetHolder: ', +timeHolderAllowance, +timeHolderWalletAllowance)

  // token = token
  //   .setAllowance(timeHolderAddress, timeHolderAllowance)
  //   .setAllowance(timeHolderWalletAddress, timeHolderWalletAllowance)
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: TIME_HOLDER_INIT, inInited: true })

  const { account } = getState().get(DUCK_SESSION)
  timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ assets, deposit ] = await Promise.all([
    timeHolderDAO.getSharesContract(),
    timeHolderDAO.getDeposit(account),
  ])
  console.log('--deposit', assets, +deposit)

  // subscribe to tokens
  const callback = (token) => dispatch(subscribeOnTokens(token))
  tokenService.on(EVENT_NEW_TOKEN, callback)
  // fetch for existing tokens
  const tokens = getState().get(DUCK_TOKENS)
  tokens.list().forEach(callback)
}

export const depositAsset = (amount: Amount, token: TokenModel) => async (dispatch) => {
  // const amountBN = new BigNumber(amount)

  // dispatch(balanceMinus(amountBN, token))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.deposit(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e.message)
  } finally {
    // compensation for update in watchTransfer
    // dispatch(balancePlus(amount, token))
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel) => async () => {
  // const amountBN = new BigNumber(amount)

  // dispatch(depositMinus(amountBN))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.withdraw(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e.message)
  } finally {
    // dispatch(depositPlus(amountBN))
  }
}
