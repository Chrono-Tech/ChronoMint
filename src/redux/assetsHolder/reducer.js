import TimeHolderModel from 'models/timeHolder/TimeHolderModel'
import * as a from './actions'

const initialState = new TimeHolderModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.TIME_HOLDER_INIT:
      return state.isInited(true)
    case a.TIME_HOLDER_TIME_ADDRESS:
      return state.timeAddress(action.address)
    case a.TIME_HOLDER_ADDRESS:
      return state.timeHolderAddress(action.address)
    case a.TIME_HOLDER_WALLET_ADDRESS:
      return state.timeHolderWalletAddress(action.address)
    default:
      return state
  }
}
