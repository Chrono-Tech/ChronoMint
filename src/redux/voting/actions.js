import Immutable from 'immutable'

import type PollModel from 'models/PollModel'
import type PollNoticeModel from 'models/notices/PollNoticeModel'

import { IS_CREATED, IS_REMOVED/*, IS_UPDATED*/ } from 'models/notices/PollNoticeModel'
import { POLLS_LOAD, POLLS_LIST } from 'redux/voting/reducer'
import { notify } from 'redux/notifier/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const watchPoll = (notice: PollNoticeModel) => async (dispatch/*, getState*/) => {
  switch (notice.status()) {
    case IS_CREATED:
      dispatch(createPoll(notice.poll))
      break
    case IS_REMOVED:
      dispatch(removePoll(notice.poll))
      break
    // case IS_UPDATED:
    //   dispatch(updatePoll(notice.poll))
    //   break
  }
  dispatch(notify(notice))
}

export const watchPolls = () => async (dispatch) => {

  const callback = (notice) => dispatch(watchPoll(notice))

  const dao = await contractsManagerDAO.getVoteDAO()

  return Promise.all([
    dao.watchCreated(callback),
    // dao.watchUpdated(callback),
    dao.watchRemoved(callback)
  ])
}

export const createPoll = (poll: PollModel) => async () => {
  console.log('createPoll', poll)
  const dao = await contractsManagerDAO.getVoteDAO()
  await dao.createPoll(poll)
}

export const removePoll = (poll: PollModel) => async () => {
  const dao = await contractsManagerDAO.getVoteDAO()
  await dao.removePoll(poll)
}

// export const updatePoll = (poll: PollModel) => async () => {
//   const dao = await contractsManagerDAO.getVoteDAO()
//   await dao.updatePoll(poll)
// }

export const loadPolls = () => async (dispatch) => {

  dispatch({ type: POLLS_LOAD })

  let list = []
  try {
    const dao = await contractsManagerDAO.getVoteDAO()
    const count = await dao.pollsCount()
    const promises = []
    for (let i = 0; i < count; i++) {
      promises.push(dao.getPoll(i))
    }
    list = Promise.all(promises)
  } finally {
    dispatch({
      type: POLLS_LIST,
      list: list.reduce((m, element) => {
        console.log('Poll item: ', element)
        return m.set()
      }, new Immutable.Map()) })
  }
}
