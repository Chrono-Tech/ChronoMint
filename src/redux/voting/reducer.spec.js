import BigNumber from 'bignumber.js'
import VotingCollection from 'models/voting/VotingCollection'
import PollModel from 'models/PollModel'
import PollDetailsModel from 'models/PollDetailsModel'
import * as a from './actions'
import reducer from './reducer'

const initialState = new VotingCollection()
const poll = new PollDetailsModel({
  poll: new PollModel()
    .set('id', 1)
    .set('published', null)
    .set('deadline', null),
})

describe('assetsManager reducer', () => {
  it('should return initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState)
  })

  it('sould POLLS_INIT', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_INIT,
        isInited: true,
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_VOTE_LIMIT', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_VOTE_LIMIT,
        voteLimitInTIME: new BigNumber(1),
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_LOAD', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_LOAD,
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_LIST', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_LIST,
        list: new VotingCollection(),
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_CREATE', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_CREATE,
        poll,
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_REMOVE_STUB', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_REMOVE_STUB,
        transactionHash: 'test',
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_UPDATE', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_UPDATE,
        poll,
      }),
    ).toMatchSnapshot()
  })

  it('sould handle POLLS_REMOVE', () => {
    expect(
      reducer(undefined, {
        type: a.POLLS_REMOVE,
        item: new PollModel().set('id', 1),
      }),
    ).toMatchSnapshot()
  })

})
