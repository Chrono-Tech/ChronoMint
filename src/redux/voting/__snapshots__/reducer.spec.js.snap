// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assetsManager reducer sould POLLS_INIT 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": true,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_CREATE 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {
    "1": {
      "files": [],
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "memberVote": null,
      "poll": {
        "active": false,
        "deadline": null,
        "description": "",
        "files": null,
        "hash": null,
        "id": 1,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isInited": false,
        "isPending": false,
        "isSelected": false,
        "isTransaction": false,
        "options": [
          "Support",
          "Decline"
        ],
        "owner": null,
        "published": null,
        "status": false,
        "timestamp": null,
        "title": "",
        "transactionHash": null,
        "voteLimitInTIME": null
      },
      "shareholdersCount": "0",
      "statistics": [],
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "votes": []
    }
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_LIST 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": true,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {
    "emptyModel": null,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isInited": false,
    "isPending": false,
    "isSelected": false,
    "leftToFetch": 0,
    "list": {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
    "voteLimitInTIME": null
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_LOAD 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": true,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_REMOVE 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_REMOVE_STUB 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_UPDATE 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {
    "1": {
      "files": [],
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isInited": false,
      "isPending": false,
      "isSelected": false,
      "memberVote": null,
      "poll": {
        "active": false,
        "deadline": null,
        "description": "",
        "files": null,
        "hash": null,
        "id": 1,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isInited": false,
        "isPending": false,
        "isSelected": false,
        "isTransaction": false,
        "options": [
          "Support",
          "Decline"
        ],
        "owner": null,
        "published": null,
        "status": false,
        "timestamp": null,
        "title": "",
        "transactionHash": null,
        "voteLimitInTIME": null
      },
      "shareholdersCount": "0",
      "statistics": [],
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "votes": []
    }
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": null
}
`;

exports[`assetsManager reducer sould handle POLLS_VOTE_LIMIT 1`] = `
{
  "emptyModel": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isInited": false,
  "isPending": false,
  "isSelected": false,
  "leftToFetch": 0,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null,
  "voteLimitInTIME": "1"
}
`;
